src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
```

Comment:
While I agree the notion of `neighbours` and `edges` feels natural from a object-oriented perspective, these are a bit problematic for the Osrank team, for a couple of reasons:

* Keeping track of the neighbours and edges _at the node level_ means more bookkeeping on our side, in a non-trivial way. The thing is, any graph library will typically store the network topology either using adjacency matrixes or adjacency lists, and therefore only the `Graph` (intended as data structure) itself will have an holistic view of things and will "know" which are the neighbours and edges for a node. This means that it might be not easy, if not not possible, to implement `neighbors` and `edges` for a concrete type, because there is no mention to the graph in this trait, and I would assume it won't be possible to access the Graph's state here, in order to implement these queries.

* Having iterators here doesn't help us, because we don't need to iterate over the neighbours -- we need to know which they are so that we can pick a new random one (by index). In a nutshell we need a bag of data here, ala `Vec<Node>` so that we can pick `neighbours[random_index]`.

* In the past I have tried playing with this idea of returning an `Iterator`, but I think in this shape it's not possible to provide a working implementation, because `Iterator` doesn't implement `Sized` and apparently if you try writing the following implementation:

```rust
fn neighbours(&self) -> ???
```

Here you are stuck: you cannot return an iterator trait object in the implementation, but if you try to return a _concrete_ iterator (e.g. `IterMut`) Rust won't accept this program, saying that the implementation doesn't match the definition. Have you tried implementing this using the `osrank` codebase? My hunch is that it won't be possible. It if is, then that's great and so much cleaner!

Reply:
Good points - I'll try to write an implementation for it next! But yes you'd have to hold a &Graph reference for this to work.

Reply:
I see what you're saying now, returning `Nodes` is the way to go.